src/utils/resource_metrics/resource_metrics.c

```diff
@@ -0,0 +1,240 @@
+/**
+ * collectd - src/utils/resource_metrics/resource_metrics.c
+ * Copyright (C) 2023       Florian octo Forster
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Florian octo Forster <octo at collectd.org>
+ **/
+
+#include "daemon/plugin.h"
+#include "utils/common/common.h"
+#include "utils/resource_metrics/resource_metrics.h"
+
+static int resource_metrics_compare(resource_metrics_t const *a,
+                                    resource_metrics_t const *b) {
+  return label_set_compare(a->resource, b->resource);
+}
+
+static resource_metrics_t *lookup_resource(resource_metrics_set_t *set,
+                                           label_set_t resource) {
+  resource_metrics_t key = {
+      .resource = resource,
+  };
+
+  return bsearch(&key, set->ptr, set->num, sizeof(*set->ptr),
+                 (void *)resource_metrics_compare);
+}
+
+static int insert_resource(resource_metrics_set_t *set, label_set_t resource) {
+  resource_metrics_t *rm =
+      realloc(set->ptr, (set->num + 1) * sizeof(*set->ptr));
+  if (rm == NULL) {
+    return ENOMEM;
+  }
+  set->ptr = rm;
+
+  rm = set->ptr + set->num;
+  memset(rm, 0, sizeof(*rm));
+
+  int status = label_set_clone(&rm->resource, resource);
+  if (status != 0) {
+    return ENOMEM;
+  }
+  set->num++;
+
+  qsort(set->ptr, set->num, sizeof(*set->ptr),
+        (void *)resource_metrics_compare);
+  return 0;
+}
+
+static resource_metrics_t *
+lookup_or_insert_resource(resource_metrics_set_t *set, label_set_t resource) {
+  resource_metrics_t *ret = lookup_resource(set, resource);
+  if (ret != NULL) {
+    return ret;
+  }
+
+  int status = insert_resource(set, resource);
+  if (status != 0) {
+    ERROR("resource_metrics: insert_resource failed: %s", STRERROR(status));
+    return NULL;
+  }
+
+  ret = lookup_resource(set, resource);
```

Comment:
Maybe insert could return the inserted pointer, to avoid need for 2nd lookup?

Reply:
Yeah, I see where you're coming from. The problem with that is that *bsearch(3)* does not tell us where in the array the new item would be expected, and so we append and run *qsort(3)*. With the information where in the array the element should go, we could implement this more efficiently.

However, that requires us to write our own binary search, do more memory management, etc. which makes the code a good deal harder to read and maintain. I think we should only go down that road after profiling has shown us that this is actually a bottleneck. "Premature optimization is the root of all evil" is one of my guiding principles. (Not that I'm consistently succeeding, there are plenty of instances in the collectd code base where I failed …)